File: app/src/main/java/com/elementary/tasks/core/utils/Sound.java
package com.elementary.tasks.core.utils;

import android.content.Context;
import android.content.res.AssetFileDescriptor;
import android.media.AudioAttributes;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.media.Ringtone;
import android.media.RingtoneManager;
import android.net.Uri;
import android.os.Handler;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.File;
import java.io.IOException;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
public class Sound {

    private Context mContext;
    @Nullable
    private MediaPlayer mMediaPlayer;
    private boolean isPaused;
    @Nullable
    private String lastFile;
    @Nullable
    private Ringtone mRingtone;
    private boolean isDone;
    @Nullable
    private PlaybackCallback mCallback;
    @NonNull
    private Handler mRingtoneHandler = new Handler();
    @NonNull
    private Runnable mRingtoneRunnable = new Runnable() {
        @Override
        public void run() {
            mRingtoneHandler.removeCallbacks(mRingtoneRunnable);
            if (mRingtone != null && mRingtone.isPlaying()) {
                mRingtoneHandler.postDelayed(mRingtoneRunnable, 100);
            } else {
                mRingtone = null;
                notifyFinish();
            }
        }
    };

    public Sound(Context context) {
        this.mContext = context;
    }

    public void setCallback(@Nullable PlaybackCallback callback) {
        this.mCallback = callback;
    }

    public void stop(boolean notify) {
        if (mMediaPlayer != null) {
            try {
                mMediaPlayer.stop();
                mMediaPlayer.release();
            } catch (IllegalStateException ignored) {
            }
            isPaused = false;
        }
        if (mRingtone != null) {
            mRingtoneHandler.removeCallbacks(mRingtoneRunnable);
            try {
                mRingtone.stop();
            } catch (Exception ignored) {
            }
        }
        if (notify) notifyFinish();
    }

    public void pause() {
        if (mMediaPlayer != null) {
            try {
                mMediaPlayer.pause();
            } catch (IllegalStateException ignored) {
            }
            isPaused = true;
        }
    }

    public void resume() {
        if (mMediaPlayer != null) {
            try {
                mMediaPlayer.start();
            } catch (IllegalStateException ignored) {
            }
            isPaused = false;
        }
    }

    public boolean isDone() {
        return isDone;
    }

    public boolean isPaused() {
        return isPaused;
    }

    public boolean isPlaying() {
        try {
            return mMediaPlayer != null && mMediaPlayer.isPlaying();
        } catch (IllegalStateException e) {
            return false;
        }
    }

    public boolean isSameFile(String path) {
        return lastFile != null && path.equalsIgnoreCase(lastFile);
    }

    public void play(String path) {
        if (!Permissions.checkPermission(mContext, Permissions.READ_EXTERNAL)) return;
        lastFile = path;
        stop(false);
        mMediaPlayer = new MediaPlayer();
        try {
            File file = new File(path);
            mMediaPlayer.setDataSource(mContext, Uri.fromFile(file));
        } catch (IOException e) {
            e.printStackTrace();
        }
        if (Module.isLollipop()) {
            AudioAttributes attributes = new AudioAttributes.Builder()
                    .setLegacyStreamType(AudioManager.STREAM_MUSIC)
                    .build();
            mMediaPlayer.setAudioAttributes(attributes);
        } else {
            mMediaPlayer.setAudioStreamType(AudioManager.STREAM_MUSIC);
        }
        mMediaPlayer.setLooping(false);
        mMediaPlayer.setOnPreparedListener(mp -> {
            notifyStart();
            mp.start();
        });
        mMediaPlayer.setOnCompletionListener(mp -> notifyFinish());
        mMediaPlayer.setOnErrorListener((mp, what, extra) -> {
            notifyFinish();
            return false;
        });
        try {
            mMediaPlayer.prepareAsync();
        } catch (IllegalStateException e) {
            e.printStackTrace();
        }
    }

    private void notifyFinish() {
        isDone = true;
        if (mCallback != null) {
            mCallback.onFinish();
        }
    }

    public void playAlarm(Uri path, boolean looping) {
        if (isPlaying() && !Permissions.checkPermission(mContext, Permissions.READ_EXTERNAL)) {
            return;
        }
        stop(false);
        mMediaPlayer = new MediaPlayer();
        try {
            mMediaPlayer.setDataSource(mContext, path);
            Prefs prefs = Prefs.getInstance(mContext);
            int stream = AudioManager.STREAM_MUSIC;
            if (prefs.isSystemLoudnessEnabled()) {
                stream = prefs.getSoundStream();
            }
            if (Module.isLollipop()) {
                AudioAttributes attributes = new AudioAttributes.Builder()
                        .setLegacyStreamType(stream)
                        .build();
                mMediaPlayer.setAudioAttributes(attributes);
            } else {
                mMediaPlayer.setAudioStreamType(stream);
            }
            mMediaPlayer.setLooping(looping);
            mMediaPlayer.setOnPreparedListener(mp -> {
                notifyStart();
                mp.start();
            });
            mMediaPlayer.setOnCompletionListener(mp -> notifyFinish());
            mMediaPlayer.setOnErrorListener((mp, what, extra) -> {
                notifyFinish();
                return false;
            });
            try {
                mMediaPlayer.prepareAsync();
            } catch (IllegalStateException e) {
                e.printStackTrace();
            }
        } catch (IOException e) {
            playRingtone(path);
        }
    }

    private void playRingtone(Uri path) {
        notifyStart();
        mRingtone = RingtoneManager.getRingtone(mContext, path);
        mRingtone.play();
        mRingtoneHandler.postDelayed(mRingtoneRunnable, 100);
    }

    public void playAlarm(AssetFileDescriptor afd) {
        if (!Permissions.checkPermission(mContext, Permissions.READ_EXTERNAL)) return;
        stop(false);
        if (isDone) {
            return;
        }
        mMediaPlayer = new MediaPlayer();
        try {
            mMediaPlayer.setDataSource(afd.getFileDescriptor(), afd.getStartOffset(), afd.getLength());
        } catch (IOException e) {
            e.printStackTrace();
        }
        Prefs prefs = Prefs.getInstance(mContext);
        int stream = AudioManager.STREAM_MUSIC;
        if (prefs.isSystemLoudnessEnabled()) {
            stream = prefs.getSoundStream();
        }
        if (Module.isLollipop()) {
            AudioAttributes attributes = new AudioAttributes.Builder()
                    .setLegacyStreamType(stream)
                    .build();
            mMediaPlayer.setAudioAttributes(attributes);
        } else {
            mMediaPlayer.setAudioStreamType(stream);
        }
        mMediaPlayer.setLooping(false);
        mMediaPlayer.setOnPreparedListener(mp -> {
            notifyStart();
            mp.start();
        });
        mMediaPlayer.setOnCompletionListener(mp -> notifyFinish());
        mMediaPlayer.setOnErrorListener((mp, what, extra) -> {
            notifyFinish();
            return false;
        });
        try {
            mMediaPlayer.prepareAsync();
        } catch (IllegalStateException e) {
            e.printStackTrace();
        }
    }

    private void notifyStart() {
        if (mCallback != null) {
            mCallback.onStart();
        }
    }

    public static boolean isDefaultMelody(String defMelody) {
        return defMelody.equals(Constants.DEFAULT);
    }

    public interface PlaybackCallback {
        void onStart();
        void onFinish();
    }
}

File: app/src/main/java/com/elementary/tasks/core/utils/Sound.kt
package com.elementary.tasks.core.utils

import android.content.Context
import android.content.res.AssetFileDescriptor
import android.media.AudioAttributes
import android.media.AudioManager
import android.media.MediaPlayer
import android.media.Ringtone
import android.media.RingtoneManager
import android.net.Uri
import android.os.Handler
import android.os.Looper
import timber.log.Timber
import java.io.File

class Sound(
  private val context: Context,
  private val prefs: Prefs
) {

  private var mediaPlayer: MediaPlayer? = null
  var isPaused: Boolean = false
    private set
  private var trimPlayback: Boolean = false
  private var playbackDuration: Int = 0
  private var playbackStartMillis: Long = 0

  private var lastFile: String? = null
  private var mRingtone: Ringtone? = null
  private var isDone: Boolean = false
  private var playbackCallback: PlaybackCallback? = null
  private val ringtoneHandler = Handler(Looper.getMainLooper())
  private val ringtoneRunnable = object : Runnable {
    override fun run() {
      ringtoneHandler.removeCallbacks(this)
      val ringtone = mRingtone
      if (ringtone != null && ringtone.isPlaying) {
        if (trimPlayback) {
          if (System.currentTimeMillis() - playbackStartMillis >= playbackDuration * 1000L) {
            stop(true)
          } else {
            ringtoneHandler.postDelayed(this, 100)
          }
        } else {
          ringtoneHandler.postDelayed(this, 100)
        }
      } else {
        stop(true)
      }
    }
  }

  private val melodyHandler = Handler(Looper.getMainLooper())
  private val melodyRunnable = object : Runnable {
    override fun run() {
      melodyHandler.removeCallbacks(this)
      val mp = mediaPlayer
      if (mp != null && mp.isPlaying) {
        if (trimPlayback) {
          if (System.currentTimeMillis() - playbackStartMillis >= playbackDuration * 1000L) {
            stop(true)
          } else {
            melodyHandler.postDelayed(this, 1000)
          }
        } else {
          melodyHandler.postDelayed(this, 1000)
        }
      } else {
        stop(true)
      }
    }
  }

  val isPlaying: Boolean
    get() {
      return try {
        val mp = mediaPlayer
        val ringtone = mRingtone
        if (mp == null && ringtone == null) return false
        if (mp != null && mp.isPlaying) return true
        if (ringtone != null && ringtone.isPlaying) return true
        false
      } catch (e: Exception) {
        false
      }
    }

  fun setCallback(callback: PlaybackCallback?) {
    this.playbackCallback = callback
  }

  fun stop(notify: Boolean) {
    ringtoneHandler.removeCallbacks(ringtoneRunnable)
    melodyHandler.removeCallbacks(melodyRunnable)
    val mp = mediaPlayer
    if (mp != null) {
      try {
        if (mp.isPlaying) {
          mp.stop()
          mp.release()
        }
      } catch (ignored: Exception) {
      }

      isPaused = false
    }
    val ringtone = mRingtone
    if (ringtone != null && ringtone.isPlaying) {
      try {
        ringtone.stop()
      } catch (ignored: Exception) {
      }
    }
    playbackStartMillis = 0L
    if (notify) {
      notifyFinish()
    }
  }

  fun pause() {
    val mp = mediaPlayer ?: return
    try {
      mp.pause()
    } catch (ignored: Exception) {
    }
    isPaused = true
  }

  fun resume() {
    val mp = mediaPlayer ?: return
    if (!isPaused) return
    try {
      mp.start()
    } catch (ignored: Exception) {
    }
    isPaused = false
  }

  fun isSameFile(path: String): Boolean {
    return lastFile != null && path.equals(lastFile!!, ignoreCase = true)
  }

  fun play(path: String) {
    if (!Permissions.checkPermission(context, Permissions.READ_EXTERNAL)) return
    lastFile = path
    stop(false)
    mediaPlayer = MediaPlayer()
    try {
      val file = File(path)
      mediaPlayer?.setDataSource(context, Uri.fromFile(file))
    } catch (e: Exception) {
      e.printStackTrace()
    }

    val attributes = AudioAttributes.Builder()
      .setLegacyStreamType(AudioManager.STREAM_MUSIC)
      .build()
    mediaPlayer?.setAudioAttributes(attributes)
    mediaPlayer?.isLooping = false
    mediaPlayer?.setOnPreparedListener { mp ->
      notifyStart()
      mp.start()
    }
    mediaPlayer?.setOnCompletionListener { notifyFinish() }
    mediaPlayer?.setOnErrorListener { _, _, _ ->
      notifyFinish()
      false
    }
    try {
      mediaPlayer?.prepareAsync()
    } catch (e: Exception) {
      e.printStackTrace()
    }
  }

  private fun notifyFinish() {
    isDone = true
    playbackCallback?.onFinish()
  }

  fun playAlarm(path: Uri, looping: Boolean, duration: Int = 0) {
    if (isPlaying || !Permissions.checkPermission(context, Permissions.READ_EXTERNAL)) {
      return
    }
    stop(false)
    trimPlayback = duration > 0 && !looping
    playbackDuration = duration
    mediaPlayer = MediaPlayer()
    try {
      mediaPlayer?.setDataSource(context, path)
      var stream = AudioManager.STREAM_MUSIC
      if (prefs.isSystemLoudnessEnabled) {
        stream = prefs.soundStream
      }
      val attributes = AudioAttributes.Builder()
        .setLegacyStreamType(stream)
        .build()
      mediaPlayer?.setAudioAttributes(attributes)
      mediaPlayer?.isLooping = looping
      mediaPlayer?.setOnPreparedListener { mp ->
        notifyStart()
        mp.start()
        playbackStartMillis = System.currentTimeMillis()
        melodyHandler.postDelayed(melodyRunnable, 1000)
      }
      mediaPlayer?.setOnCompletionListener { notifyFinish() }
      mediaPlayer?.setOnErrorListener { _, _, _ ->
        notifyFinish()
        false
      }
      mediaPlayer?.prepareAsync()
    } catch (e: Exception) {
      playRingtone(path)
    }
  }

  fun playRingtone(path: Uri) {
    Timber.d("playRingtone: $path")
    notifyStart()
    mRingtone = RingtoneManager.getRingtone(context, path)
    mRingtone?.play()
    playbackStartMillis = System.currentTimeMillis()
    ringtoneHandler.postDelayed(ringtoneRunnable, 100)
  }

  fun playAlarm(afd: AssetFileDescriptor) {
    if (!Permissions.checkPermission(context, Permissions.READ_EXTERNAL)) return
    stop(false)
    if (isDone) {
      return
    }
    trimPlayback = false
    mediaPlayer = MediaPlayer()
    try {
      mediaPlayer?.setDataSource(afd.fileDescriptor, afd.startOffset, afd.length)
    } catch (e: Exception) {
      e.printStackTrace()
    }

    var stream = AudioManager.STREAM_MUSIC
    if (prefs.isSystemLoudnessEnabled) {
      stream = prefs.soundStream
    }
    val attributes = AudioAttributes.Builder()
      .setLegacyStreamType(stream)
      .build()
    mediaPlayer?.setAudioAttributes(attributes)
    mediaPlayer?.isLooping = false
    mediaPlayer?.setOnPreparedListener { mp ->
      notifyStart()
      mp.start()
    }
    mediaPlayer?.setOnCompletionListener { notifyFinish() }
    mediaPlayer?.setOnErrorListener { _, _, _ ->
      notifyFinish()
      false
    }
    try {
      mediaPlayer?.prepareAsync()
    } catch (e: Exception) {
      e.printStackTrace()
    }
  }

  private fun notifyStart() {
    playbackCallback?.onStart()
  }

  interface PlaybackCallback {
    fun onStart()
    fun onFinish()
  }

  companion object {

    fun isDefaultMelody(defMelody: String): Boolean {
      return defMelody == Constants.SOUND_ALARM || defMelody == Constants.SOUND_NOTIFICATION
        || defMelody == Constants.SOUND_RINGTONE || defMelody == Constants.DEFAULT
    }
  }
}
